eclipse-theia/theia
# Add Electron start-up performance script

#### What it does

Adds a new performance script for measurement of frontend start-up performance of the Electron example, similar to what was done in PR #9777. In fact, much of that script is factored out into a module of reusable functions to build the Electron script. This new script runs in much the same way, with similar command-line arguments to specify the number of runs and where to put the Chromium trace files.

Resolves #10440.

Example output:

```console
[Performance][Electron Frontend Startup][1] Largest Contentful Paint (LCP): 6.716 seconds
[Performance][Electron Frontend Startup][2] Largest Contentful Paint (LCP): 6.037 seconds
[Performance][Electron Frontend Startup][3] Largest Contentful Paint (LCP): 5.992 seconds
[Performance][Electron Frontend Startup][4] Largest Contentful Paint (LCP): 6.026 seconds
[Performance][Electron Frontend Startup][5] Largest Contentful Paint (LCP): 6.052 seconds
[Performance][Electron Frontend Startup][6] Largest Contentful Paint (LCP): 6.137 seconds
[Performance][Electron Frontend Startup][7] Largest Contentful Paint (LCP): 6.081 seconds
[Performance][Electron Frontend Startup][8] Largest Contentful Paint (LCP): 5.964 seconds
[Performance][Electron Frontend Startup][9] Largest Contentful Paint (LCP): 5.779 seconds
[Performance][Electron Frontend Startup][10] Largest Contentful Paint (LCP): 6.147 seconds
[Performance][Electron Frontend Startup][MEAN] Largest Contentful Paint (LCP): 6.093 seconds
[Performance][Electron Frontend Startup][STDEV] Largest Contentful Paint (LCP): 0.230 seconds
```

Note that the existing NPM script for the browser start-up performance is renamed from `performance:startup` to `performance:startup:browser` to distinguish it from the Electron case.

A new command-line option `--app` (`-a`) is added to the `extension-impact.js` script that selects the example application to build for execution of the tests. Supported values are `browser` and `electron`; the default if not specified is the browser app as before.

No new package dependencies are added.

Contributed on behalf of STMicroelectronics.

#### How to test

See the readme in the `scripts/performance/` directory for details of how to run the script. For most cases, it is sufficient just to run the NPM script from the monorepo root:

```console
$ yarn performance:startup:electron
```

This will print out performance measurements with summary statistics as shown above. You may also import the traces generated by the script in the `scripts/performance/profiles/electron/` directory into the Performance tab of the Chrome Developer Tools for analysis.

Also, try running the Extension Impact script with the new option to drive the Electron app:

```console
$ cd scripts/performance
$ node ./extension-performance.js --app electron
```

**Note** that executing the extension impact script on the Electron app after having run it on the browser app will often fail to get measurements for the `@theia/git` extension; it throws an exception about a mismatch in node versions in the native modules:

```console
Error: The module '/home/runner/work/theia/theia/node_modules/find-git-repositories/build/Release/findGitRepos.node'
was compiled against a different Node.js version using NODE_MODULE_VERSION 83. This version of Node.js requires
NODE_MODULE_VERSION 80. Please try re-compiling or re-installing the module (for instance, using `npm rebuild` or
`npm install`).
```

This can be remedied by cleaning the Git workspace and building it again before running the script.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
